pkg/controller/master/rancher/embedded_rancher.go

```diff
@@ -30,3 +36,67 @@ func (h *Handler) RancherSettingOnChange(key string, setting *rancherv3api.Setti
 	}
 	return nil, nil
 }
+
+// initializeTLS writes internal-cacerts to cacerts value and adds the VIP to certificate SANs
+// cacerts is used in generated kubeconfig files
+func (h *Handler) initializeTLS(setting *rancherv3api.Setting) error {
+	cacerts, err := h.RancherSettingCache.Get("cacerts")
+	if err != nil {
+		return err
+	}
+	if cacerts.Value != "" {
+		return nil
+	}
+
+	if err := h.setCACerts(setting.Value); err != nil {
+		return err
+	}
+
+	if err := h.addVIPToSAN(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (h *Handler) setCACerts(cert string) error {
+	cacerts, err := h.RancherSettingCache.Get("cacerts")
+	if err != nil {
+		return err
+	}
+	if cacerts.Value == cert {
```

Comment:
should this be `if cacerts.Value == cert || cacerts.Value != ""`

Reply:
The `cacerts.Value != ""` check is used to skip the update when it is already initialized. But for a function like `setCACerts(cert)`, it doesn't have to care about that.